In [27]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


In [28]:
df = pd.read_csv('/kaggle/input/nasa-asteroids-classification/nasa.csv')

In [29]:
df.head()

Neo Reference ID     Name  Absolute Magnitude  Est Dia in KM(min)  \
0           3703080  3703080                21.6            0.127220   
1           3723955  3723955                21.3            0.146068   
2           2446862  2446862                20.3            0.231502   
3           3092506  3092506                27.4            0.008801   
4           3514799  3514799                21.6            0.127220   

   Est Dia in KM(max)  Est Dia in M(min)  Est Dia in M(max)  \
0            0.284472         127.219879         284.472297   
1            0.326618         146.067964         326.617897   
2            0.517654         231.502122         517.654482   
3            0.019681           8.801465          19.680675   
4            0.284472         127.219879         284.472297   

   Est Dia in Miles(min)  Est Dia in Miles(max)  Est Dia in Feet(min)  ...  \
0               0.079051               0.176763            417.388066  ...   
1               0.090762               0.202951            479.225620  ...   
2               0.143849               0.321655            759.521423  ...   
3               0.005469               0.012229             28.876199  ...   
4               0.079051               0.176763            417.388066  ...   

   Asc Node Longitude Orbital Period  Perihelion Distance  Perihelion Arg  \
0          314.373913     609.599786             0.808259       57.257470   
1          136.717242     425.869294             0.718200      313.091975   
2          259.475979     643.580228             0.950791      248.415038   
3           57.173266     514.082140             0.983902       18.707701   
4           84.629307     495.597821             0.967687      158.263596   

   Aphelion Dist  Perihelion Time  Mean Anomaly  Mean Motion  Equinox  \
0       2.005764     2.458162e+06    264.837533     0.590551    J2000   
1       1.497352     2.457795e+06    173.741112     0.845330    J2000   
2       1.966857     2.458120e+06    292.893654     0.559371    J2000   
3       1.527904     2.457902e+06     68.741007     0.700277    J2000   
4       1.483543     2.457814e+06    135.142133     0.726395    J2000   

   Hazardous  
0       True  
1      False  
2       True  
3      False  
4       True  

[5 rows x 40 columns]

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4687 entries, 0 to 4686
Data columns (total 40 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Neo Reference ID              4687 non-null   int64  
 1   Name                          4687 non-null   int64  
 2   Absolute Magnitude            4687 non-null   float64
 3   Est Dia in KM(min)            4687 non-null   float64
 4   Est Dia in KM(max)            4687 non-null   float64
 5   Est Dia in M(min)             4687 non-null   float64
 6   Est Dia in M(max)             4687 non-null   float64
 7   Est Dia in Miles(min)         4687 non-null   float64
 8   Est Dia in Miles(max)         4687 non-null   float64
 9   Est Dia in Feet(min)          4687 non-null   float64
 10  Est Dia in Feet(max)          4687 non-null   float64
 11  Close Approach Date           4687 non-null   object 
 12  Epoch Date Close Approach     4687 non-null   int64  
 13  Rel

In [31]:
df.isnull().sum()

Neo Reference ID                0
Name                            0
Absolute Magnitude              0
Est Dia in KM(min)              0
Est Dia in KM(max)              0
Est Dia in M(min)               0
Est Dia in M(max)               0
Est Dia in Miles(min)           0
Est Dia in Miles(max)           0
Est Dia in Feet(min)            0
Est Dia in Feet(max)            0
Close Approach Date             0
Epoch Date Close Approach       0
Relative Velocity km per sec    0
Relative Velocity km per hr     0
Miles per hour                  0
Miss Dist.(Astronomical)        0
Miss Dist.(lunar)               0
Miss Dist.(kilometers)          0
Miss Dist.(miles)               0
Orbiting Body                   0
Orbit ID                        0
Orbit Determination Date        0
Orbit Uncertainity              0
Minimum Orbit Intersection      0
Jupiter Tisserand Invariant     0
Epoch Osculation                0
Eccentricity                    0
Semi Major Axis                 0
Inclination   

In [32]:
df.duplicated().sum()

0

# Deleting Unnecessary columns

In [33]:
df.drop(columns=['Neo Reference ID','Name','Est Dia in KM(max)','Est Dia in M(min)','Est Dia in M(max)','Est Dia in Miles(min)','Est Dia in Miles(max)','Est Dia in Feet(min)','Est Dia in Feet(max)',
'Close Approach Date','Epoch Date Close Approach','Relative Velocity km per hr','Miss Dist.(Astronomical)','Miss Dist.(lunar)','Miss Dist.(miles)','Orbit ID',
],inplace=True,axis=1)

In [34]:
df.head()

Absolute Magnitude  Est Dia in KM(min)  Relative Velocity km per sec  \
0                21.6            0.127220                      6.115834   
1                21.3            0.146068                     18.113985   
2                20.3            0.231502                      7.590711   
3                27.4            0.008801                     11.173874   
4                21.6            0.127220                      9.840831   

   Miles per hour  Miss Dist.(kilometers) Orbiting Body  \
0    13680.509944              62753692.0         Earth   
1    40519.173105              57298148.0         Earth   
2    16979.661798               7622911.5         Earth   
3    24994.839864              42683616.0         Earth   
4    22012.954985              61010824.0         Earth   

  Orbit Determination Date  Orbit Uncertainity  Minimum Orbit Intersection  \
0      2017-04-06 08:36:37                   5                    0.025282   
1      2017-04-06 08:32:49                   3                    0.186935   
2      2017-04-06 09:20:19                   0                    0.043058   
3      2017-04-06 09:15:49                   6                    0.005512   
4      2017-04-06 08:57:58                   1                    0.034798   

   Jupiter Tisserand Invariant  ...  Asc Node Longitude  Orbital Period  \
0                        4.634  ...          314.373913      609.599786   
1                        5.457  ...          136.717242      425.869294   
2                        4.557  ...          259.475979      643.580228   
3                        5.093  ...           57.173266      514.082140   
4                        5.154  ...           84.629307      495.597821   

   Perihelion Distance  Perihelion Arg  Aphelion Dist  Perihelion Time  \
0             0.808259       57.257470       2.005764     2.458162e+06   
1             0.718200      313.091975       1.497352     2.457795e+06   
2             0.950791      248.415038       1.966857     2.458120e+06   
3             0.983902       18.707701       1.527904     2.457902e+06   
4             0.967687      158.263596       1.483543     2.457814e+06   

   Mean Anomaly  Mean Motion  Equinox  Hazardous  
0    264.837533     0.590551    J2000       True  
1    173.741112     0.845330    J2000      False  
2    292.893654     0.559371    J2000       True  
3     68.741007     0.700277    J2000      False  
4    135.142133     0.726395    J2000       True  

[5 rows x 24 columns]

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4687 entries, 0 to 4686
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Absolute Magnitude            4687 non-null   float64
 1   Est Dia in KM(min)            4687 non-null   float64
 2   Relative Velocity km per sec  4687 non-null   float64
 3   Miles per hour                4687 non-null   float64
 4   Miss Dist.(kilometers)        4687 non-null   float64
 5   Orbiting Body                 4687 non-null   object 
 6   Orbit Determination Date      4687 non-null   object 
 7   Orbit Uncertainity            4687 non-null   int64  
 8   Minimum Orbit Intersection    4687 non-null   float64
 9   Jupiter Tisserand Invariant   4687 non-null   float64
 10  Epoch Osculation              4687 non-null   float64
 11  Eccentricity                  4687 non-null   float64
 12  Semi Major Axis               4687 non-null   float64
 13  Inc

In [36]:
#checkin orbitting body of asteroids
df['Orbiting Body'].value_counts()

Orbiting Body
Earth    4687
Name: count, dtype: int64

:-All asteroid orbitring around only earth,so we can drop this column as it is not giving any relation with hazarsdous or not

In [37]:
#checking Equinox column values
df['Equinox'].value_counts()

Equinox
J2000    4687
Name: count, dtype: int64

-:Here Equinox of all the asteroids are same ,so we can ignore this column.

Dropping more three unnecessary column

In [38]:
df.drop(columns=['Equinox','Orbiting Body','Orbit Determination Date'],inplace=True,axis=1)

In [39]:
df.head()

Absolute Magnitude  Est Dia in KM(min)  Relative Velocity km per sec  \
0                21.6            0.127220                      6.115834   
1                21.3            0.146068                     18.113985   
2                20.3            0.231502                      7.590711   
3                27.4            0.008801                     11.173874   
4                21.6            0.127220                      9.840831   

   Miles per hour  Miss Dist.(kilometers)  Orbit Uncertainity  \
0    13680.509944              62753692.0                   5   
1    40519.173105              57298148.0                   3   
2    16979.661798               7622911.5                   0   
3    24994.839864              42683616.0                   6   
4    22012.954985              61010824.0                   1   

   Minimum Orbit Intersection  Jupiter Tisserand Invariant  Epoch Osculation  \
0                    0.025282                        4.634         2458000.5   
1                    0.186935                        5.457         2458000.5   
2                    0.043058                        4.557         2458000.5   
3                    0.005512                        5.093         2458000.5   
4                    0.034798                        5.154         2458000.5   

   Eccentricity  ...  Inclination  Asc Node Longitude  Orbital Period  \
0      0.425549  ...     6.025981          314.373913      609.599786   
1      0.351674  ...    28.412996          136.717242      425.869294   
2      0.348248  ...     4.237961          259.475979      643.580228   
3      0.216578  ...     7.905894           57.173266      514.082140   
4      0.210448  ...    16.793382           84.629307      495.597821   

   Perihelion Distance  Perihelion Arg  Aphelion Dist  Perihelion Time  \
0             0.808259       57.257470       2.005764     2.458162e+06   
1             0.718200      313.091975       1.497352     2.457795e+06   
2             0.950791      248.415038       1.966857     2.458120e+06   
3             0.983902       18.707701       1.527904     2.457902e+06   
4             0.967687      158.263596       1.483543     2.457814e+06   

   Mean Anomaly  Mean Motion  Hazardous  
0    264.837533     0.590551       True  
1    173.741112     0.845330      False  
2    292.893654     0.559371       True  
3     68.741007     0.700277      False  
4    135.142133     0.726395       True  

[5 rows x 21 columns]

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4687 entries, 0 to 4686
Data columns (total 21 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Absolute Magnitude            4687 non-null   float64
 1   Est Dia in KM(min)            4687 non-null   float64
 2   Relative Velocity km per sec  4687 non-null   float64
 3   Miles per hour                4687 non-null   float64
 4   Miss Dist.(kilometers)        4687 non-null   float64
 5   Orbit Uncertainity            4687 non-null   int64  
 6   Minimum Orbit Intersection    4687 non-null   float64
 7   Jupiter Tisserand Invariant   4687 non-null   float64
 8   Epoch Osculation              4687 non-null   float64
 9   Eccentricity                  4687 non-null   float64
 10  Semi Major Axis               4687 non-null   float64
 11  Inclination                   4687 non-null   float64
 12  Asc Node Longitude            4687 non-null   float64
 13  Orb

# Distribution of Our data

In [41]:
# import seaborn as sns
# import matplotlib.pyplot as plt

# def plot_numerical_columns(df):
#     numerical_columns = df.select_dtypes(include=['int64', 'float64']).columns
    
#     # Plot histograms and boxplots side by side for each numerical column
#     fig, axes = plt.subplots(nrows=len(numerical_columns), ncols=2, figsize=(12, 6*len(numerical_columns)))
    
#     for i, column in enumerate(numerical_columns):
#         # Histogram plot
#         sns.histplot(df[column], ax=axes[i, 0])
#         axes[i, 0].set_title(f'Histogram of {column}')
        
#         # Boxplot plot
#         sns.boxplot(x=df[column], ax=axes[i, 1])
#         axes[i, 1].set_title(f'Boxplot of {column}')
        
#     plt.tight_layout()
#     plt.show()

# # Example usage:
# # Assuming 'df' is your DataFrame containing numerical columns
# plot_numerical_columns(df)


In [42]:
df.head()

Absolute Magnitude  Est Dia in KM(min)  Relative Velocity km per sec  \
0                21.6            0.127220                      6.115834   
1                21.3            0.146068                     18.113985   
2                20.3            0.231502                      7.590711   
3                27.4            0.008801                     11.173874   
4                21.6            0.127220                      9.840831   

   Miles per hour  Miss Dist.(kilometers)  Orbit Uncertainity  \
0    13680.509944              62753692.0                   5   
1    40519.173105              57298148.0                   3   
2    16979.661798               7622911.5                   0   
3    24994.839864              42683616.0                   6   
4    22012.954985              61010824.0                   1   

   Minimum Orbit Intersection  Jupiter Tisserand Invariant  Epoch Osculation  \
0                    0.025282                        4.634         2458000.5   
1                    0.186935                        5.457         2458000.5   
2                    0.043058                        4.557         2458000.5   
3                    0.005512                        5.093         2458000.5   
4                    0.034798                        5.154         2458000.5   

   Eccentricity  ...  Inclination  Asc Node Longitude  Orbital Period  \
0      0.425549  ...     6.025981          314.373913      609.599786   
1      0.351674  ...    28.412996          136.717242      425.869294   
2      0.348248  ...     4.237961          259.475979      643.580228   
3      0.216578  ...     7.905894           57.173266      514.082140   
4      0.210448  ...    16.793382           84.629307      495.597821   

   Perihelion Distance  Perihelion Arg  Aphelion Dist  Perihelion Time  \
0             0.808259       57.257470       2.005764     2.458162e+06   
1             0.718200      313.091975       1.497352     2.457795e+06   
2             0.950791      248.415038       1.966857     2.458120e+06   
3             0.983902       18.707701       1.527904     2.457902e+06   
4             0.967687      158.263596       1.483543     2.457814e+06   

   Mean Anomaly  Mean Motion  Hazardous  
0    264.837533     0.590551       True  
1    173.741112     0.845330      False  
2    292.893654     0.559371       True  
3     68.741007     0.700277      False  
4    135.142133     0.726395       True  

[5 rows x 21 columns]

# Checking target coloumn

In [43]:
df['Hazardous'].value_counts()

Hazardous
False    3932
True      755
Name: count, dtype: int64

:- We got to know that it is highly imbalanced column ,having false  value 3932 and true value 755.
I think we have to do SMOTE,we will do it later

# Encoding hazardous and non-hazardous asteroids

In [44]:
from sklearn.preprocessing import LabelEncoder,MinMaxScaler
le = LabelEncoder()
df["Hazardous"] = le.fit_transform(df["Hazardous"])

# Scaling selected features

In [45]:
X = df.drop('Hazardous', axis=1)
y = df['Hazardous']

scaler = MinMaxScaler()
X = scaler.fit_transform(X)

In [46]:
x_copy = df.drop('Hazardous', axis=1)


In [47]:
x_copy.head()

Absolute Magnitude  Est Dia in KM(min)  Relative Velocity km per sec  \
0                21.6            0.127220                      6.115834   
1                21.3            0.146068                     18.113985   
2                20.3            0.231502                      7.590711   
3                27.4            0.008801                     11.173874   
4                21.6            0.127220                      9.840831   

   Miles per hour  Miss Dist.(kilometers)  Orbit Uncertainity  \
0    13680.509944              62753692.0                   5   
1    40519.173105              57298148.0                   3   
2    16979.661798               7622911.5                   0   
3    24994.839864              42683616.0                   6   
4    22012.954985              61010824.0                   1   

   Minimum Orbit Intersection  Jupiter Tisserand Invariant  Epoch Osculation  \
0                    0.025282                        4.634         2458000.5   
1                    0.186935                        5.457         2458000.5   
2                    0.043058                        4.557         2458000.5   
3                    0.005512                        5.093         2458000.5   
4                    0.034798                        5.154         2458000.5   

   Eccentricity  Semi Major Axis  Inclination  Asc Node Longitude  \
0      0.425549         1.407011     6.025981          314.373913   
1      0.351674         1.107776    28.412996          136.717242   
2      0.348248         1.458824     4.237961          259.475979   
3      0.216578         1.255903     7.905894           57.173266   
4      0.210448         1.225615    16.793382           84.629307   

   Orbital Period  Perihelion Distance  Perihelion Arg  Aphelion Dist  \
0      609.599786             0.808259       57.257470       2.005764   
1      425.869294             0.718200      313.091975       1.497352   
2      643.580228             0.950791      248.415038       1.966857   
3      514.082140             0.983902       18.707701       1.527904   
4      495.597821             0.967687      158.263596       1.483543   

   Perihelion Time  Mean Anomaly  Mean Motion  
0     2.458162e+06    264.837533     0.590551  
1     2.457795e+06    173.741112     0.845330  
2     2.458120e+06    292.893654     0.559371  
3     2.457902e+06     68.741007     0.700277  
4     2.457814e+06    135.142133     0.726395

# SMOTE OF HAZARDOUS COLUMN

In [48]:
from imblearn.over_sampling import SMOTE




# SMOTE object
smote = SMOTE(sampling_strategy='auto', random_state=42)

# Resample the data
X_resampled, y_resampled = smote.fit_resample(X, y)


In [49]:
y_resampled.value_counts()

Hazardous
1    3932
0    3932
Name: count, dtype: int64

In [50]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=1)


# Feature importances - which feature affects whether asteroid is hazardous or not

In [51]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from xgboost import XGBClassifier

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [52]:
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)

# Fit the classifier to the training data
rf_classifier.fit(X_train, y_train)

# Get feature importances
importances = rf_classifier.feature_importances_

# Create dummy feature names
feature_names = x_copy.columns

# Sort feature importances in descending order
indices = np.argsort(importances)[::-1]

# Plot the feature importances with column names
plt.figure(figsize=(10, 6))
plt.title("Feature Importances")
plt.bar(range(X_train.shape[1]), importances[indices], align="center")
plt.xticks(range(X_train.shape[1]), [feature_names[i] for i in indices], rotation=90)  # Set column names as x-axis ticks
plt.xlabel("Feature")
plt.ylabel("Importance Score")
plt.tight_layout()  # Adjust layout to prevent overlap of labels
plt.show()

NameError: name 'plt' is not defined

In [ ]:
# rfc = RandomForestClassifier()
# abc = AdaBoostClassifier()
# gbc = GradientBoostingClassifier()
# etc = ExtraTreesClassifier()
# lgr = LogisticRegression()
# svc = SVC()
# dtc = DecisionTreeClassifier()
# xgb = XGBClassifier()

# models = [rfc, abc, gbc, etc, lgr, svc, dtc, xgb]
# names = ["Random Forest", "Ada Boost", "Gradient Boosting",
#         "Extra Trees", "Logistic Regression", "Support Vectors",
#         "Decision Tree", "XGBoost"]

In [53]:
# def training(model, name, axes):
#     model.fit(X_train, y_train)
#     pred = model.predict(X_test)
#     score = accuracy_score(y_test, pred)
#     report = classification_report(y_test, pred)
#     cm = confusion_matrix(y_test, pred)
#     sns.heatmap(cm, annot=True, ax=axes)
#     axes.set_title("{}: {}%".format(name, round(score*100, 2)))
#     return score*100, report

In [54]:
# scores = []
# reports = dict()
# index = 0
# fig, axes = plt.subplots(nrows=2, ncols=4, figsize=(10, 7))
# for i in range(2):
#     for j in range(4):
#         score, report = training(models[index], names[index], axes[i][j])
#         scores += [score]
#         reports[names[index]] = report
#         index += 1
# plt.tight_layout()
# plt.show()

In [60]:
y_pred

Hazardous
0    796
1    777
Name: count, dtype: int64

# GridSearcchCV

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
param_grid = {
    'n_estimators': [50,100,120],
    'max_depth': [None,4,5,8],
    'min_samples_split': [2, 3],
    'min_samples_leaf': [1, 2],
    'max_features': [1.0, 'sqrt']
}
rf = RandomForestRegressor()
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)
print("Best Parameters:", grid_search.best_params_)
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)
score = accuracy_score(y_test, y_pred)


Fitting 5 folds for each of 96 candidates, totalling 480 fits


In [58]:
print(score)

NameError: name 'score' is not defined

[CV] END max_depth=5, max_features=1.0, min_samples_leaf=2, min_samples_split=2, n_estimators=50; total time=   1.0s
[CV] END max_depth=5, max_features=1.0, min_samples_leaf=2, min_samples_split=2, n_estimators=100; total time=   2.2s
[CV] END max_depth=5, max_features=1.0, min_samples_leaf=2, min_samples_split=2, n_estimators=120; total time=   2.3s
[CV] END max_depth=5, max_features=1.0, min_samples_leaf=2, min_samples_split=2, n_estimators=120; total time=   2.3s
[CV] END max_depth=5, max_features=1.0, min_samples_leaf=2, min_samples_split=3, n_estimators=100; total time=   2.0s
[CV] END max_depth=5, max_features=1.0, min_samples_leaf=2, min_samples_split=3, n_estimators=120; total time=   2.3s
[CV] END max_depth=5, max_features=1.0, min_samples_leaf=2, min_samples_split=3, n_estimators=120; total time=   2.3s
[CV] END max_depth=5, max_features=sqrt, min_samples_leaf=1, min_samples_split=2, n_estimators=120; total time=   1.0s
[CV] END max_depth=5, max_features=sqrt, min_samples_lea

In [57]:
best_model

RandomForestRegressor()